File: anvil-constraintlayout/src/main/java/trikita/anvil/constraint/layout/ConstraintDSL.java
package trikita.anvil.constraint.layout;

import android.view.View;

import androidx.constraintlayout.widget.Barrier;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.constraintlayout.widget.ConstraintSet;
import androidx.constraintlayout.widget.Group;
import androidx.constraintlayout.widget.Guideline;
import androidx.constraintlayout.widget.Placeholder;

import java.util.LinkedList;

import trikita.anvil.Anvil;
import trikita.anvil.BaseDSL;

/**
 * DSL for creating views and settings their attributes.
 * It contains views and their setters from the library constraint-layout.
 */
public final class ConstraintDSL implements Anvil.AttributeSetter {
    static {
        Anvil.registerAttributeSetter(new ConstraintDSL());
    }

    public static BaseDSL.ViewClassResult constraintLayout() {
        return BaseDSL.v(ConstraintLayout.class);
    }

    public static Void constraintLayout(final Anvil.Renderable r) {
        return BaseDSL.v(ConstraintLayout.class, new Anvil.Renderable() {
            @Override
            public void view() {
                r.view();
                applyConstraints();
            }
        });
    }

    public static BaseDSL.ViewClassResult barrier() {
        return BaseDSL.v(Barrier.class);
    }

    public static Void barrier(Anvil.Renderable r) {
        return BaseDSL.v(Barrier.class, r);
    }

    public static BaseDSL.ViewClassResult placeholder() {
        return BaseDSL.v(Placeholder.class);
    }

    public static Void placeholder(Anvil.Renderable r) {
        return BaseDSL.v(Placeholder.class, r);
    }

    public static BaseDSL.ViewClassResult group() {
        return BaseDSL.v(Group.class);
    }

    public static Void group(Anvil.Renderable r) {
        return BaseDSL.v(Group.class, r);
    }

    public static BaseDSL.ViewClassResult guideline() {
        return BaseDSL.v(Guideline.class);
    }

    public static Void guideline(Anvil.Renderable r) {
        return BaseDSL.v(Guideline.class, r);
    }

    public static Void type(int arg) {
        return BaseDSL.attr("type", arg);
    }

    public static Void referencedIds(int[] arg) {
        return BaseDSL.attr("referencedIds", arg);
    }

    public static Void leftConstraintToLeft(int toId) {
        return BaseDSL.attr("leftConstraintToLeft", toId);
    }

    public static Void leftConstraintToRight(int toId) {
        return BaseDSL.attr("leftConstraintToRight", toId);
    }

    public static Void leftConstraintToTop(int toId) {
        return BaseDSL.attr("leftConstraintToTop", toId);
    }

    public static Void leftConstraintToBottom(int toId) {
        return BaseDSL.attr("leftConstraintToBottom", toId);
    }

    public static Void leftConstraintToParent() {
        return leftConstraintToLeft(ConstraintSet.PARENT_ID);
    }

    public static Void rightConstraintToRight(int toId) {
        return BaseDSL.attr("rightConstraintToRight", toId);
    }

    public static Void rightConstraintToLeft(int toId) {
        return BaseDSL.attr("rightConstraintToLeft", toId);
    }

    public static Void rightConstraintToTop(int toId) {
        return BaseDSL.attr("rightConstraintToTop", toId);
    }

    public static Void rightConstraintToBottom(int toId) {
        return BaseDSL.attr("rightConstraintToBottom", toId);
    }

    public static Void rightConstraintToParent() {
        return rightConstraintToRight(ConstraintSet.PARENT_ID);
    }

    public static Void topConstraintToRight(int toId) {
        return BaseDSL.attr("topConstraintToRight", toId);
    }

    public static Void topConstraintToLeft(int toId) {
        return BaseDSL.attr("topConstraintToLeft", toId);
    }

    public static Void topConstraintToTop(int toId) {
        return BaseDSL.attr("topConstraintToTop", toId);
    }

    public static Void topConstraintToBottom(int toId) {
        return BaseDSL.attr("topConstraintToBottom", toId);
    }

    public static Void topConstraintToParent() {
        return topConstraintToTop(ConstraintSet.PARENT_ID);
    }

    public static Void bottomConstraintToRight(int toId) {
        return BaseDSL.attr("bottomConstraintToRight", toId);
    }

    public static Void bottomConstraintToLeft(int toId) {
        return BaseDSL.attr("bottomConstraintToLeft", toId);
    }

    public static Void bottomConstraintToTop(int toId) {
        return BaseDSL.attr("bottomConstraintToTop", toId);
    }

    public static Void bottomConstraintToBottom(int toId) {
        return BaseDSL.attr("bottomConstraintToBottom", toId);
    }

    public static Void bottomConstraintToParent() {
        return bottomConstraintToBottom(ConstraintSet.PARENT_ID);
    }

    public static Void chain(int leftId, int leftSide, int rightId, int rightSide, int[] chainIds, float[] weights, int style) {
        return BaseDSL.attr("chain", new ConstraintChain(leftId, leftSide, rightId, rightSide, chainIds, weights, style));
    }

    public static Void circleConstraint(int id, int angle, int radius) {
        return BaseDSL.attr("circleConstraint", new CircleConstraint(0, id, angle, radius));
    }

    public static Void id(int id) {
        return BaseDSL.attr("idForConstraint", id);
    }

    public static Void applyConstraints() {
        return BaseDSL.attr("apply", true);
    }

    public static Void orientation(int orientation) {
        return BaseDSL.attr("orientation", orientation);
    }

    public static Void guideBegin(int margin) {
        return BaseDSL.attr("guideBegin", margin);
    }

    public static Void guideEnd(int margin) {
        return BaseDSL.attr("guideEnd", margin);
    }

    public static Void guidelinePercent(float ratio) {
        return BaseDSL.attr("guidelinePercent", ratio);
    }

    public static Void contentId(int contentId) {
        return BaseDSL.attr("contentId", contentId);
    }

    public final int CONSTRAINTS_KEY = 1 + 2 << 24;
    public final int CHAIN_KEY = 1 + 3 << 24;
    public final int CIRCLE_KEY = 1 + 4 << 24;

    public boolean set(View v, String name, final Object arg, final Object old) {
        switch (name) {
            case "allowsGoneWidget":
                if (v instanceof Barrier && arg instanceof Boolean) {
                    ((Barrier) v).setAllowsGoneWidget((boolean) arg);
                    return true;
                }
                break;
            case "constraintSet":
                if (v instanceof ConstraintLayout && arg instanceof ConstraintSet) {
                    ((ConstraintLayout) v).setConstraintSet((ConstraintSet) arg);
                    return true;
                }
                break;
            case "leftConstraintToLeft":
                if (v.getParent() instanceof ConstraintLayout && arg instanceof Integer) {
                    addConstraint(((ConstraintLayout) v.getParent()), v.getId(), ConstraintSet.LEFT, (Integer) arg, ConstraintSet.LEFT);
                    return true;
                }
                break;
            case "leftConstraintToRight":
                if (v.getParent() instanceof ConstraintLayout && arg instanceof Integer) {
                    addConstraint(((ConstraintLayout) v.getParent()), v.getId(), ConstraintSet.LEFT, (Integer) arg, ConstraintSet.RIGHT);
                    return true;
                }
                break;
            case "leftConstraintToTop":
                if (v.getParent() instanceof ConstraintLayout && arg instanceof Integer) {
                    addConstraint(((ConstraintLayout) v.getParent()), v.getId(), ConstraintSet.LEFT, (Integer) arg, ConstraintSet.TOP);
                    return true;
                }
                break;
            case "leftConstraintToBottom":
                if (v.getParent() instanceof ConstraintLayout && arg instanceof Integer) {
                    addConstraint(((ConstraintLayout) v.getParent()), v.getId(), ConstraintSet.LEFT, (Integer) arg, ConstraintSet.BOTTOM);
                    return true;
                }
                break;
            case "rightConstraintToLeft":
                if (v.getParent() instanceof ConstraintLayout && arg instanceof Integer) {
                    addConstraint(((ConstraintLayout) v.getParent()), v.getId(), ConstraintSet.RIGHT, (Integer) arg, ConstraintSet.LEFT);
                    return true;
                }
                break;
            case "rightConstraintToRight":
                if (v.getParent() instanceof ConstraintLayout && arg instanceof Integer) {
                    addConstraint(((ConstraintLayout) v.getParent()), v.getId(), ConstraintSet.RIGHT, (Integer) arg, ConstraintSet.RIGHT);
                    return true;
                }
                break;
            case "rightConstraintToTop":
                if (v.getParent() instanceof ConstraintLayout && arg instanceof Integer) {
                    addConstraint(((ConstraintLayout) v.getParent()), v.getId(), ConstraintSet.RIGHT, (Integer) arg, ConstraintSet.TOP);
                    return true;
                }
                break;
            case "rightConstraintToBottom":
                if (v.getParent() instanceof ConstraintLayout && arg instanceof Integer) {
                    addConstraint(((ConstraintLayout) v.getParent()), v.getId(), ConstraintSet.RIGHT, (Integer) arg, ConstraintSet.BOTTOM);
                    return true;
                }
                break;
            case "topConstraintToLeft":
                if (v.getParent() instanceof ConstraintLayout && arg instanceof Integer) {
                    addConstraint(((ConstraintLayout) v.getParent()), v.getId(), ConstraintSet.TOP, (Integer) arg, ConstraintSet.LEFT);
                    return true;
                }
                break;
            case "topConstraintToRight":
                if (v.getParent() instanceof ConstraintLayout && arg instanceof Integer) {
                    addConstraint(((ConstraintLayout) v.getParent()), v.getId(), ConstraintSet.TOP, (Integer) arg, ConstraintSet.RIGHT);
                    return true;
                }
                break;
            case "topConstraintToTop":
                if (v.getParent() instanceof ConstraintLayout && arg instanceof Integer) {
                    addConstraint(((ConstraintLayout) v.getParent()), v.getId(), ConstraintSet.TOP, (Integer) arg, ConstraintSet.TOP);
                    return true;
                }
                break;
            case "topConstraintToBottom":
                if (v.getParent() instanceof ConstraintLayout && arg instanceof Integer) {
                    addConstraint(((ConstraintLayout) v.getParent()), v.getId(), ConstraintSet.TOP, (Integer) arg, ConstraintSet.BOTTOM);
                    return true;
                }
                break;
            case "bottomConstraintToLeft":
                if (v.getParent() instanceof ConstraintLayout && arg instanceof Integer) {
                    addConstraint(((ConstraintLayout) v.getParent()), v.getId(), ConstraintSet.BOTTOM, (Integer) arg, ConstraintSet.LEFT);
                    return true;
                }
                break;
            case "bottomConstraintToRight":
                if (v.getParent() instanceof ConstraintLayout && arg instanceof Integer) {
                    addConstraint(((ConstraintLayout) v.getParent()), v.getId(), ConstraintSet.BOTTOM, (Integer) arg, ConstraintSet.RIGHT);
                    return true;
                }
                break;
            case "bottomConstraintToTop":
                if (v.getParent() instanceof ConstraintLayout && arg instanceof Integer) {
                    addConstraint(((ConstraintLayout) v.getParent()), v.getId(), ConstraintSet.BOTTOM, (Integer) arg, ConstraintSet.TOP);
                    return true;
                }
                break;
            case "bottomConstraintToBottom":
                if (v.getParent() instanceof ConstraintLayout && arg instanceof Integer) {
                    addConstraint(((ConstraintLayout) v.getParent()), v.getId(), ConstraintSet.BOTTOM, (Integer) arg, ConstraintSet.BOTTOM);
                    return true;
                }
                break;
            case "chain":
                if (v.getParent() instanceof ConstraintLayout && arg instanceof ConstraintChain) {
                    ConstraintLayout cl = ((ConstraintLayout) v.getParent());
                    LinkedList<ConstraintChain> chainsList = (LinkedList<ConstraintChain>) cl.getTag(CHAIN_KEY);
                    if (chainsList == null) {
                        chainsList = new LinkedList<>();
                    }
                    chainsList.add((ConstraintChain) arg);
                    cl.setTag(CHAIN_KEY, chainsList);
                    return true;
                }
                break;
            case "circleConstraint":
                if (v.getParent() instanceof ConstraintLayout && arg instanceof CircleConstraint) {
                    ConstraintLayout cl = ((ConstraintLayout) v.getParent());
                    LinkedList<CircleConstraint> circlesList = (LinkedList<CircleConstraint>) cl.getTag(CIRCLE_KEY);
                    if (circlesList == null) {
                        circlesList = new LinkedList<>();
                    }
                    CircleConstraint arg2 = (CircleConstraint) arg;
                    arg2.centerId = v.getId();
                    circlesList.add(arg2);
                    cl.setTag(CIRCLE_KEY, circlesList);
                    return true;
                }
                break;
            case "apply":
                if (v instanceof ConstraintLayout && arg instanceof Boolean) {
                    ConstraintLayout cl = ((ConstraintLayout) v);
                    LinkedList<ConstraintSide> constraintSideList = (LinkedList<ConstraintSide>) cl.getTag(CONSTRAINTS_KEY);
                    LinkedList<ConstraintChain> chainsList = (LinkedList<ConstraintChain>) cl.getTag(CHAIN_KEY);
                    LinkedList<CircleConstraint> circlesList = (LinkedList<CircleConstraint>) cl.getTag(CIRCLE_KEY);

                    if ((constraintSideList == null || constraintSideList.size() == 0)
                            && (chainsList == null || chainsList.size() == 0)
                            && (circlesList == null || circlesList.size() == 0)) {
                        return false;
                    }

                    ConstraintSet cs1 = new ConstraintSet();
                    cs1.clone(cl);

                    if (constraintSideList != null && constraintSideList.size() > 0) {
                        for (int i = 0; i < constraintSideList.size(); i++) {
                            ConstraintSide side = constraintSideList.get(i);
                            cs1.connect(side.startID, side.startSide, side.endID, side.endSide);
                        }
                        constraintSideList.clear();
                    }

                    if (chainsList != null && chainsList.size() > 0) {
                        for (int i = 0; i < chainsList.size(); i++) {
                            ConstraintChain chain = chainsList.get(i);
                            cs1.createHorizontalChain(chain.leftId, chain.leftSide, chain.rightId, chain.rightSide, chain.chainIds, chain.weights, chain.style);
                        }
                        chainsList.clear();
                    }

                    if (circlesList != null && circlesList.size() > 0) {
                        for (int i = 0; i < circlesList.size(); i++) {
                            CircleConstraint circleConstraint = circlesList.get(i);
                            cs1.constrainCircle(circleConstraint.id, circleConstraint.centerId, circleConstraint.radius, circleConstraint.angle);
                        }
                        circlesList.clear();
                    }

                    cs1.applyTo(cl);
                    return false;
                }
                break;
            case "maxHeight":
                if (v instanceof ConstraintLayout && arg instanceof Integer) {
                    ((ConstraintLayout) v).setMaxHeight((int) arg);
                    return true;
                }
                break;
            case "maxWidth":
                if (v instanceof ConstraintLayout && arg instanceof Integer) {
                    ((ConstraintLayout) v).setMaxWidth((int) arg);
                    return true;
                }
                break;
            case "minHeight":
                if (v instanceof ConstraintLayout && arg instanceof Integer) {
                    ((ConstraintLayout) v).setMinHeight((int) arg);
                    return true;
                }
                break;
            case "minWidth":
                if (v instanceof ConstraintLayout && arg instanceof Integer) {
                    ((ConstraintLayout) v).setMinWidth((int) arg);
                    return true;
                }
                break;
            case "optimizationLevel":
                if (v instanceof ConstraintLayout && arg instanceof Integer) {
                    ((ConstraintLayout) v).setOptimizationLevel((int) arg);
                    return true;
                }
                break;
            case "referencedIds":
                if (v instanceof Barrier && arg instanceof int[]) {
                    ((Barrier) v).setReferencedIds((int[]) arg);
                    return true;
                }
                if (v instanceof Group && arg instanceof int[]) {
                    ((Group) v).setReferencedIds((int[]) arg);
                    return true;
                }
                break;
            case "type":
                if (v instanceof Barrier && arg instanceof Integer) {
                    ((Barrier) v).setType((int) arg);
                    return true;
                }
                break;
            case "idForConstraint":
                if (arg instanceof Integer) {
                    v.setId((Integer) arg);
                    if (v.getParent() instanceof ConstraintLayout) {
                        ((ConstraintLayout) v.getParent()).onViewAdded(v);
                    }
                    return true;
                }
                break;
            case "orientation":
                if (v instanceof Guideline && arg instanceof Integer) {
                    ConstraintLayout.LayoutParams params = (ConstraintLayout.LayoutParams) v.getLayoutParams();
                    params.orientation = (Integer) arg;
                    v.setLayoutParams(params);
                    return true;
                }
                break;
            case "guideBegin":
                if (v instanceof Guideline && arg instanceof Integer) {
                    ConstraintLayout.LayoutParams params = (ConstraintLayout.LayoutParams) v.getLayoutParams();
                    params.guideBegin = (Integer) arg;
                    v.setLayoutParams(params);
                    return true;
                }
                break;
            case "guideEnd":
                if (v instanceof Guideline && arg instanceof Integer) {
                    ConstraintLayout.LayoutParams params = (ConstraintLayout.LayoutParams) v.getLayoutParams();
                    params.guideEnd = (Integer) arg;
                    v.setLayoutParams(params);
                    return true;
                }
                break;
            case "guidelinePercent":
                if (v instanceof Guideline && arg instanceof Float) {
                    ConstraintLayout.LayoutParams params = (ConstraintLayout.LayoutParams) v.getLayoutParams();
                    params.guidePercent = (Float) arg;
                    v.setLayoutParams(params);
                    return true;
                }
                break;
            case "contentId":
                if (v instanceof Placeholder && arg instanceof Integer) {
                    ((Placeholder) v).setContentId((int) arg);
                    return true;
                }
                break;
        }
        return false;
    }

    private void addConstraint(ConstraintLayout cl, int fromId, int fromSide, int toId, int endSide) {
        LinkedList<ConstraintSide> constraintSideList = (LinkedList<ConstraintSide>) cl.getTag(CONSTRAINTS_KEY);
        if (constraintSideList == null) {
            constraintSideList = new LinkedList<>();
            cl.setTag(CONSTRAINTS_KEY, constraintSideList);
        }
        constraintSideList.add(
                new ConstraintSide(fromId, fromSide, toId, endSide));
    }
}

File: anvil-constraintlayout/src/main/kotlin/trikita/anvil/ConstraintDsl.kt
@file:Suppress("DEPRECATION", "UNCHECKED_CAST", "MemberVisibilityCanBePrivate", "unused")

package trikita.anvil

import android.view.View
import androidx.constraintlayout.widget.*
import java.util.*

/*
 * TODO many of these methods should be available for children of ConstraintLayout but not ConstraintLayout itself.
 *   Revisit this when multi-receivers are available.
*/

fun constraintLayout(configure: ConstraintLayoutScope.() -> Unit = {}) =
    v<ConstraintLayout, ConstraintLayoutScope>(ConstraintLayoutScope) {
        configure()
        applyConstraints()
    }
abstract class ConstraintLayoutScope : ViewGroupScope() {
    fun sideConstraint(toId: Int, fromSide: Int, toSide: Int): Unit =
        attr("sideConstraint", ConstraintSide(0, fromSide, toId, toSide))
    fun leftConstraintToLeft(toId: Int): Unit = sideConstraint(toId, ConstraintSet.LEFT, ConstraintSet.LEFT)
    fun leftConstraintToRight(toId: Int): Unit = sideConstraint(toId, ConstraintSet.LEFT, ConstraintSet.RIGHT)
    fun leftConstraintToTop(toId: Int): Unit = sideConstraint(toId, ConstraintSet.LEFT, ConstraintSet.TOP)
    fun leftConstraintToBottom(toId: Int): Unit = sideConstraint(toId, ConstraintSet.LEFT, ConstraintSet.BOTTOM)
    fun leftConstraintToParent(): Unit = leftConstraintToLeft(ConstraintSet.PARENT_ID)
    fun rightConstraintToLeft(toId: Int): Unit = sideConstraint(toId, ConstraintSet.RIGHT, ConstraintSet.LEFT)
    fun rightConstraintToRight(toId: Int): Unit = sideConstraint(toId, ConstraintSet.RIGHT, ConstraintSet.RIGHT)
    fun rightConstraintToTop(toId: Int): Unit = sideConstraint(toId, ConstraintSet.RIGHT, ConstraintSet.TOP)
    fun rightConstraintToBottom(toId: Int): Unit = sideConstraint(toId, ConstraintSet.RIGHT, ConstraintSet.BOTTOM)
    fun rightConstraintToParent(): Unit = rightConstraintToRight(ConstraintSet.PARENT_ID)
    fun topConstraintToLeft(toId: Int): Unit = sideConstraint(toId, ConstraintSet.TOP, ConstraintSet.LEFT)
    fun topConstraintToRight(toId: Int): Unit = sideConstraint(toId, ConstraintSet.TOP, ConstraintSet.RIGHT)
    fun topConstraintToTop(toId: Int): Unit = sideConstraint(toId, ConstraintSet.TOP, ConstraintSet.TOP)
    fun topConstraintToBottom(toId: Int): Unit = sideConstraint(toId, ConstraintSet.TOP, ConstraintSet.BOTTOM)
    fun topConstraintToParent(): Unit = topConstraintToTop(ConstraintSet.PARENT_ID)
    fun bottomConstraintToLeft(toId: Int): Unit = sideConstraint(toId, ConstraintSet.BOTTOM, ConstraintSet.LEFT)
    fun bottomConstraintToRight(toId: Int): Unit = sideConstraint(toId, ConstraintSet.BOTTOM, ConstraintSet.RIGHT)
    fun bottomConstraintToTop(toId: Int): Unit = sideConstraint(toId, ConstraintSet.BOTTOM, ConstraintSet.TOP)
    fun bottomConstraintToBottom(toId: Int): Unit = sideConstraint(toId, ConstraintSet.BOTTOM, ConstraintSet.BOTTOM)
    fun bottomConstraintToParent(): Unit = bottomConstraintToBottom(ConstraintSet.PARENT_ID)
    fun horizontalChain(
        leftId: Int, leftSide: Int, rightId: Int, rightSide: Int,
        chainIds: IntArray?, weights: FloatArray?, style: Int): Unit = attr(
        "chainConstraint",
        ConstraintChain(true, leftId, leftSide, rightId, rightSide, chainIds, weights, style))
    fun verticalChain(
        leftId: Int, leftSide: Int, rightId: Int, rightSide: Int,
        chainIds: IntArray?, weights: FloatArray?, style: Int): Unit = attr(
        "chainConstraint",
        ConstraintChain(false, leftId, leftSide, rightId, rightSide, chainIds, weights, style))
    fun circleConstraint(id: Int, angle: Int, radius: Int): Unit =
        attr("circleConstraint", ConstraintCircle(0, id, angle, radius))
    fun constraintId(arg: Int): Unit = attr("constraintId", arg)
    fun applyConstraints(): Unit = attr("applyConstraints", Unit)

    companion object : ConstraintLayoutScope() {
        init {
            Anvil.registerAttributeSetter(ConstraintDslSetter)}
    }
}

abstract class ConstraintHelperScope : ViewScope() {
    fun referencedIds(vararg arg: Int): Unit = attr("referencedIds", arg)
}

fun barrier(configure: BarrierScope.() -> Unit = {}) =
    v<Barrier>(configure.bind(BarrierScope))
abstract class BarrierScope : ConstraintHelperScope() {
    fun type(arg: Int): Unit = attr("type", arg)
    fun allowsGoneWidget(arg: Boolean): Unit = attr("allowsGoneWidget", arg)
    companion object : BarrierScope()
}

fun group(configure: GroupScope.() -> Unit = {}) =
    v<Group>(configure.bind(GroupScope))
abstract class GroupScope : ConstraintHelperScope() {
    companion object : GroupScope()
}

fun placeholder(configure: PlaceholderScope.() -> Unit = {}) =
    v<Placeholder>(configure.bind(PlaceholderScope))
abstract class PlaceholderScope : ViewGroupScope() {
    fun contentId(contentId: Int): Unit = attr("contentId", contentId)
    fun emptyVisibility(visibility: Int): Unit = attr("emptyVisibility", visibility)
    companion object : PlaceholderScope()
}

fun guideline(configure: GuidelineScope.() -> Unit = {}) =
    v<Guideline>(configure.bind(GuidelineScope))
abstract class GuidelineScope : ViewGroupScope() {
    fun orientation(orientation: Int): Unit = attr("orientation", orientation)
    fun guideBegin(margin: Int): Unit = attr("guideBegin", margin)
    fun guideEnd(margin: Int): Unit = attr("guideEnd", margin)
    fun guidelinePercent(ratio: Float): Unit = attr("guidelinePercent", ratio)
    companion object : GuidelineScope()
}

object ConstraintDslSetter : Anvil.AttributeSetter<Any?> {
    override fun set(
        v: View,
        name: String,
        arg: Any?,
        old: Any?
    ): Boolean = when (name) {
        "constraintSet" -> when {
            v is ConstraintLayout && arg is ConstraintSet -> {
                v.setConstraintSet(arg)
                true
            }
            else -> false
        }
        "sideConstraint" -> when {
            v.parent is ConstraintLayout && arg is ConstraintSide -> {
                (v.parent as ConstraintLayout).taggedSides += arg.apply { startId = v.id }
                true
            }
            else -> false
        }
        "chainConstraint" -> when {
            v.parent is ConstraintLayout && arg is ConstraintChain -> {
                (v.parent as ConstraintLayout).taggedChains += arg
                true
            }
            else -> false
        }
        "circleConstraint" -> when {
            v.parent is ConstraintLayout && arg is ConstraintCircle -> {
                (v.parent as ConstraintLayout).taggedCircles += arg.apply { centerId = v.id }
                true
            }
            else -> false
        }
        "applyConstraints" -> when {
            v is ConstraintLayout -> {
                val sides = v.taggedSidesOrNull
                val chains = v.taggedChainsOrNull
                val circles = v.taggedCirclesOrNull
                if (!sides.isNullOrEmpty() || !chains.isNullOrEmpty() || !circles.isNullOrEmpty()) {
                    v.updateConstraints {
                        sides?.let {
                            it.forEach { side ->
                                connect(
                                    side.startId,
                                    side.startSide,
                                    side.endId,
                                    side.endSide
                                )
                            }
                            it.clear()
                        }

                        chains?.let {
                            it.forEach { chain ->
                                if(chain.isHorizontal) {
                                    createHorizontalChain(
                                        chain.leftId,
                                        chain.leftSide,
                                        chain.rightId,
                                        chain.rightSide,
                                        chain.chainIds,
                                        chain.weights,
                                        chain.style
                                    )
                                } else {
                                    createVerticalChain(
                                        chain.leftId,
                                        chain.leftSide,
                                        chain.rightId,
                                        chain.rightSide,
                                        chain.chainIds,
                                        chain.weights,
                                        chain.style
                                    )
                                }
                            }
                            it.clear()
                        }

                        circles?.let {
                            it.forEach { circle ->
                                constrainCircle(
                                    circle.id,
                                    circle.centerId,
                                    circle.radius,
                                    circle.angle.toFloat()
                                )
                            }
                            it.clear()
                        }
                    }
                    false
                } else false
            }
            else -> false
        }
        "maxHeight" -> when {
            v is ConstraintLayout && arg is Int -> {
                v.maxHeight = arg
                true
            }
            else -> false
        }
        "maxWidth" -> when {
            v is ConstraintLayout && arg is Int -> {
                v.maxWidth = arg
                true
            }
            else -> false
        }
        "minHeight" -> when {
            v is ConstraintLayout && arg is Int -> {
                v.minHeight = arg
                true
            }
            else -> false
        }
        "minWidth" -> when {
            v is ConstraintLayout && arg is Int -> {
                v.minWidth = arg
                true
            }
            else -> false
        }
        "optimizationLevel" -> when {
            v is ConstraintLayout && arg is Int -> {
                v.optimizationLevel = arg
                true
            }
            else -> false
        }
        "constraintId" -> when {
            arg is Int -> {
                v.id = arg
                if (v.parent is ConstraintLayout) {
                    (v.parent as ConstraintLayout).onViewAdded(v)
                }
                true
            }
            else -> false
        }
        "referencedIds" -> when {
            v is ConstraintHelper && arg is IntArray -> {
                v.referencedIds = arg
                true
            }
            else -> false
        }
        "type" -> when {
            v is Barrier && arg is Int -> {
                v.type = arg
                true
            }
            else -> false
        }
        "allowsGoneWidget" -> when {
            v is Barrier && arg is Boolean -> {
                v.setAllowsGoneWidget(arg)
                true
            }
            else -> false
        }
        "contentId" -> when {
            v is Placeholder && arg is Int -> {
                v.setContentId(arg)
                true
            }
            else -> false
        }
        "emptyVisibility" -> when {
            v is Placeholder && arg is Int -> {
                v.emptyVisibility = arg
                true
            }
            else -> false
        }
        "orientation" -> when {
            v is Guideline && arg is Int -> {
                val params = v.getLayoutParams() as ConstraintLayout.LayoutParams
                params.orientation = arg
                v.setLayoutParams(params)
                true
            }
            else -> false
        }
        "guideBegin" -> when {
            v is Guideline && arg is Int -> {
                val params = v.getLayoutParams() as ConstraintLayout.LayoutParams
                params.guideBegin = arg
                v.setLayoutParams(params)
                true
            }
            else -> false
        }
        "guideEnd" -> when {
            v is Guideline && arg is Int -> {
                val params = v.getLayoutParams() as ConstraintLayout.LayoutParams
                params.guideEnd = arg
                v.setLayoutParams(params)
                true
            }
            else -> false
        }
        "guidelinePercent" -> when {
            v is Guideline && arg is Float -> {
                val params = v.getLayoutParams() as ConstraintLayout.LayoutParams
                params.guidePercent = arg
                v.setLayoutParams(params)
                true
            }
            else -> false
        }
        else -> false
    }
}

private const val CONSTRAINTS_KEY: Int = 1 + 2 shl 24
private const val CHAIN_KEY = 1 + 3 shl 24
private const val CIRCLE_KEY = 1 + 4 shl 24

private val ConstraintLayout.taggedSidesOrNull: MutableList<ConstraintSide>?
    get() = getTag(CONSTRAINTS_KEY) as? LinkedList<ConstraintSide>

private val ConstraintLayout.taggedSides: MutableList<ConstraintSide>
    get() = taggedSidesOrNull ?: LinkedList<ConstraintSide>().also { setTag(CONSTRAINTS_KEY, it) }

private val ConstraintLayout.taggedChainsOrNull: MutableList<ConstraintChain>?
    get() = getTag(CHAIN_KEY) as? LinkedList<ConstraintChain>

private val ConstraintLayout.taggedChains: MutableList<ConstraintChain>
    get() = taggedChainsOrNull ?: LinkedList<ConstraintChain>().also { setTag(CHAIN_KEY, it) }

private val ConstraintLayout.taggedCirclesOrNull: MutableList<ConstraintCircle>?
    get() = getTag(CIRCLE_KEY) as? LinkedList<ConstraintCircle>

private val ConstraintLayout.taggedCircles: MutableList<ConstraintCircle>
    get() = taggedCirclesOrNull ?: LinkedList<ConstraintCircle>().also { setTag(CIRCLE_KEY, it) }

private data class ConstraintSide(
    var startId: Int,
    val startSide: Int,
    val endId: Int,
    val endSide: Int
)

private data class ConstraintCircle(
    var centerId: Int,
    val id: Int,
    val angle: Int,
    val radius: Int
)

private data class ConstraintChain(
    val isHorizontal: Boolean,
    val leftId: Int,
    val leftSide: Int,
    val rightId: Int,
    val rightSide: Int,
    val chainIds: IntArray?,
    val weights: FloatArray?,
    val style: Int
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ConstraintChain

        if (isHorizontal != other.isHorizontal) return false
        if (leftId != other.leftId) return false
        if (leftSide != other.leftSide) return false
        if (rightId != other.rightId) return false
        if (rightSide != other.rightSide) return false
        if (chainIds != null) {
            if (other.chainIds == null) return false
            if (!chainIds.contentEquals(other.chainIds)) return false
        } else if (other.chainIds != null) return false
        if (weights != null) {
            if (other.weights == null) return false
            if (!weights.contentEquals(other.weights)) return false
        } else if (other.weights != null) return false
        if (style != other.style) return false

        return true
    }

    override fun hashCode(): Int {
        var result = leftId
        result = 31 * result + if(isHorizontal) 1 else 0
        result = 31 * result + leftSide
        result = 31 * result + rightId
        result = 31 * result + rightSide
        result = 31 * result + (chainIds?.contentHashCode() ?: 0)
        result = 31 * result + (weights?.contentHashCode() ?: 0)
        result = 31 * result + style
        return result
    }
}

private fun ConstraintLayout.updateConstraints(action: ConstraintSet.() -> Unit) {
    val cs = ConstraintSet()
    cs.clone(this)
    action(cs)
    cs.applyTo(this)
}
